### PR TITLE
New Resource and Data Source: aws_acmpca_certificate_authority

### LINKED ISSUE #4059: Feature Request: Support AWS Certificate Manager Private Certificate Authorities
Amazon has released a new service, AWS Certificate Manager Private Certificate Authority (ACM PCA), which is a managed CA service. https://aws.amazon.com/about-aws/whats-new/2018/04/introducing-aws-certificate-manager-private-certificate-authority/

Please vote on this issue by adding a :+1: reaction to the original issue to help prioritize interest. If you're interested in implementing this feature request, please comment below.

Prerequisites:
* aws-sdk-go v1.13.28: #4048
* `meta.(*AWSClient).acmpcaconn`: #4049

### Terraform Version

terraform 0.10+

### New Resource(s)

- `aws_acmpca_certificate_authority`

### Terraform Configuration Files

```hcl
# Example implementation, may change when developed
resource "aws_acmpca_certificate_authority" "example" {
  certificate_authority_configuration { # Required
  	key_algorithm     = ""
  	signing_algorithm = ""

  	subject {
      common_name                 = ""
      country                     = ""
      distinguised_name_qualifier = ""
      generation_qualifier        = ""
      given_name                  = ""
      initials                    = ""
      locality                    = ""
      organization                = ""
      organizational_unit         = ""
      pseudonym                   = ""
      state                       = ""
      surname                     = ""
      title                       = ""
  	}
  }
  
  revocation_configuration { # Optional
    crl_configuration {
      custom_cname       = ""
      enabled            = true
      expiration_in_days = 1
      s3_bucket_name     = "${aws_s3_bucket.example.name}"
    }
  }
}
```

### Expected Behavior

Resource creates, updates, and deletes ACM private certificate authorities.

### Actual Behavior

New feature.

### References

* https://docs.aws.amazon.com/acm-pca/latest/userguide/
* https://docs.aws.amazon.com/sdk-for-go/api/service/acmpca/#ACMPCA.CreateCertificateAuthority

### PR DESCRIPTION
**Pull request by bflad**
Closes #4059 

```
make testacc TEST=./aws TESTARGS='-run=TestAccAwsAcmpcaCertificateAuthority'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAwsAcmpcaCertificateAuthority -timeout 120m
=== RUN   TestAccAwsAcmpcaCertificateAuthority_Basic
--- PASS: TestAccAwsAcmpcaCertificateAuthority_Basic (18.96s)
=== RUN   TestAccAwsAcmpcaCertificateAuthority_Enabled
--- SKIP: TestAccAwsAcmpcaCertificateAuthority_Enabled (0.00s)
	resource_aws_acmpca_certificate_authority_test.go:106: We need to fully sign the certificate authority CSR from another CA in order to test this functionality, which requires another resource
=== RUN   TestAccAwsAcmpcaCertificateAuthority_RevocationConfiguration_CrlConfiguration_CustomCname
--- PASS: TestAccAwsAcmpcaCertificateAuthority_RevocationConfiguration_CrlConfiguration_CustomCname (112.38s)
=== RUN   TestAccAwsAcmpcaCertificateAuthority_RevocationConfiguration_CrlConfiguration_Enabled
--- PASS: TestAccAwsAcmpcaCertificateAuthority_RevocationConfiguration_CrlConfiguration_Enabled (89.18s)
=== RUN   TestAccAwsAcmpcaCertificateAuthority_RevocationConfiguration_CrlConfiguration_ExpirationInDays
--- PASS: TestAccAwsAcmpcaCertificateAuthority_RevocationConfiguration_CrlConfiguration_ExpirationInDays (67.04s)
=== RUN   TestAccAwsAcmpcaCertificateAuthority_Tags
--- PASS: TestAccAwsAcmpcaCertificateAuthority_Tags (51.20s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	338.792s

make testacc TEST=./aws TESTARGS='-run=TestAccDataSourceAwsAcmpcaCertificateAuthority_Basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccDataSourceAwsAcmpcaCertificateAuthority_Basic -timeout 120m
=== RUN   TestAccDataSourceAwsAcmpcaCertificateAuthority_Basic
--- PASS: TestAccDataSourceAwsAcmpcaCertificateAuthority_Basic (19.96s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	19.997s
```

### CODE REVIEW

File: aws/resource_aws_acmpca_certificate_authority.go [hunk 1]
```diff
@@ -0,0 +1,736 @@
...
+			return nil
+		}
+		// Returned when in PENDING_CERTIFICATE status
+		// InvalidStateException: The certificate authority XXXXX is not in the correct state to have a certificate signing request.
+		if !isAWSErr(err, acmpca.ErrCodeInvalidStateException, "") {
+			return fmt.Errorf("error reading ACMPCA Certificate Authority Certificate: %s", err)
+		}
+	}
+
+	d.Set("certificate", "")
```
Comment (mbfrahry) [inline aws/resource_aws_acmpca_certificate_authority.go]: I haven't seen this before. What's the reasoning emptying this value?

File: website/docs/r/acmpca_certificate_authority.html.markdown [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+* `enabled` - (Optional) Whether the certificate authority is enabled or disabled. Defaults to `true`.
+* `revocation_configuration` - (Optional) Nested argument containing revocation configuration. Defined below.
+* `tags` - (Optional) Specifies a key-value map of user-defined tags that are attached to the certificate authority.
+* `type` - (Optional) The type of the certificate authority. Currently, this must be `SUBORDINATE`.
+
+### certificate_authority_configuration
+
+* `key_algorithm` - (Required) Type of the public key algorithm and size, in bits, of the key pair that your key pair creates when it issues a certificate. Valid values can be found in the [ACM PCA Documentation](https://docs.aws.amazon.com/acm-pca/latest/APIReference/API_CertificateAuthorityConfiguration.html).
+* `signing_algorithm` - (Required) Name of the algorithm your private CA uses to sign certificate requests. Valid values can be found in the [ACM PCA Documentation](https://docs.aws.amazon.com/acm-pca/latest/APIReference/API_CertificateAuthorityConfiguration.html).
+* `subject` - (Required) Nested argument that contains X.500 distinguished name information
```
Comment (mbfrahry) [inline website/docs/r/acmpca_certificate_authority.html.markdown]: The attributes for `subject` are all optional but `subject` itself is required. Are one of these attributes required when you create this resource?

Comment (bflad) [reply website/docs/r/acmpca_certificate_authority.html.markdown]: Ha yup, I missed adding that bit to the documentation. Great catch!

```
		* aws_acmpca_certificate_authority.test: error creating ACMPCA Certificate Authority: ValidationException: 1 validation error detected: The subject must have at least one attribute set.
			status code: 400, request id: afff1cd3-5b85-4d7e-abc3-06a14aecc2c7
```

**Comment by bflad** [comment]
Update passes TravisCI and acceptance testing, merging!

```
make testacc TEST=./aws TESTARGS='-run=TestAccAwsAcmpcaCertificateAuthority'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAwsAcmpcaCertificateAuthority -timeout 120m
=== RUN   TestAccAwsAcmpcaCertificateAuthority_Basic
--- PASS: TestAccAwsAcmpcaCertificateAuthority_Basic (19.15s)
=== RUN   TestAccAwsAcmpcaCertificateAuthority_Enabled
--- SKIP: TestAccAwsAcmpcaCertificateAuthority_Enabled (0.00s)
	resource_aws_acmpca_certificate_authority_test.go:106: We need to fully sign the certificate authority CSR from another CA in order to test this functionality, which requires another resource
=== RUN   TestAccAwsAcmpcaCertificateAuthority_RevocationConfiguration_CrlConfiguration_CustomCname
--- PASS: TestAccAwsAcmpcaCertificateAuthority_RevocationConfiguration_CrlConfiguration_CustomCname (113.12s)
=== RUN   TestAccAwsAcmpcaCertificateAuthority_RevocationConfiguration_CrlConfiguration_Enabled
--- PASS: TestAccAwsAcmpcaCertificateAuthority_RevocationConfiguration_CrlConfiguration_Enabled (88.83s)
=== RUN   TestAccAwsAcmpcaCertificateAuthority_RevocationConfiguration_CrlConfiguration_ExpirationInDays
--- PASS: TestAccAwsAcmpcaCertificateAuthority_RevocationConfiguration_CrlConfiguration_ExpirationInDays (78.24s)
=== RUN   TestAccAwsAcmpcaCertificateAuthority_Tags
--- PASS: TestAccAwsAcmpcaCertificateAuthority_Tags (53.03s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	352.410s
```

**Comment by bflad** [comment]
This has been released in [version 1.18.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1180-may-10-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**